### PR TITLE
Feature/28 a method that returns information about all wallets

### PR DESCRIPTION
**Pull request by pyanik**

### CODE REVIEW

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletDTOMapper.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package pl.byczazagroda.trackexpensesappbackend.service;
+
+import org.springframework.stereotype.Service;
+import pl.byczazagroda.trackexpensesappbackend.dto.WalletDTO;
+import pl.byczazagroda.trackexpensesappbackend.model.Wallet;
+
+import java.time.Instant;
+
+@Service
+class WalletDTOMapper {
```
Comment (dawciobiel) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletDTOMapper.java:10]: In this project we are going to use MapStruct. So you don't have to implement it yourself.

**Review by dawciobiel — APPROVED** [review]

**Review by dawciobiel — CHANGES_REQUESTED** [review]
Changes are required.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java [hunk 2]
```diff
@@ -14,4 +22,21 @@
...
     private final WalletService walletService;
+
+    private final WalletRepository walletRepository;
+
+    @GetMapping(value = "/wallets")
+    ResponseEntity<List<WalletOutput>> getWallets() throws AppException {
+        try {
+            List<WalletOutput> allWallets = walletService.getAllWallets();
+
+            if (allWallets.isEmpty())
```
[the remark this answers is not in the document]
Comment (pyanik) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java]: Brackets have been added

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/dto/WalletDTO.java [hunk 3]
```diff
@@ -9,5 +11,20 @@
  * @param name is a Wallet's name
  * @param creationDate is a date and hour when wallet has been created
  */
-public record WalletDTO(Long id, String name, Instant creationDate) {
+public record WalletDTO(Long id, String name, Instant creationDate) implements WalletOutput {
```
[the remark this answers is not in the document]
Comment (pyanik) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/dto/WalletDTO.java]: WalletOutput class removed

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletOutput.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package pl.byczazagroda.trackexpensesappbackend.service;
+
+import java.time.Instant;
+
+/**
+ * Interface for passing the output data related to the wallet entity
```
[the remark this answers is not in the document]
Comment (pyanik) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletOutput.java]: Class removed with accordance to another issue

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletDTOMapper.java (hunk 1 above)
Comment (pyanik) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletDTOMapper.java]: Custom WalletDTOMapper class has been replaced with MapStruct

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java [hunk 5]
```diff
@@ -1,13 +1,30 @@
...
 
+    private final WalletModelMapper walletModelMapper;
+
+    @Override
+    public List<WalletDTO> getAllWallets() {
+        log.info("Return list of all wallets");
+        return walletRepository.findAll()
+                .stream()
+                .map(walletModelMapper::mapWalletEntityToWalletDTO)
+                .collect(Collectors.toList());
```
Comment (Norsu296) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java:28]: Samo .toList wystarczy, nie ma potrzeby dawać Collectors

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java [hunk 6]
```diff
@@ -14,4 +21,20 @@
...
+        try {
+            List<WalletDTO> allWallets = walletService.getAllWallets();
+
+            if (allWallets.isEmpty()) {
+                return ResponseEntity.status(HttpStatus.NO_CONTENT).body(allWallets);
+            }
+
+            return ResponseEntity.status(HttpStatus.OK).body(allWallets);
+
+        } catch (Exception e) {
```
Comment (Norsu296) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java:36]: Po co ten try catch? Nie powinno być to zrobione w serwisie? Jak już ten try catch to poleciał error i co dalej, pasuje jakiś log dodać

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java [hunk 7]
```diff
@@ -14,4 +21,20 @@
 public class WalletController {
 
     private final WalletService walletService;
+
+    @GetMapping(value = "/wallets")
```
Comment (Norsu296) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java:25]: Niepotrzebne to value, nad klasą adnotacja @RequestMapping to załatwi, jak będzie więcej metod to po co w każdej znowu value dodawać

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java [hunk 8]
```diff
@@ -14,4 +21,20 @@
...
+    ResponseEntity<List<WalletDTO>> getWallets() throws AppException {
```
Comment (Norsu296) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java:26]: Po co to response entity? Całą logikę bym dał do serwisu i w controllerze tylko wywołanie

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java (hunk 5 above)
Comment (pyanik) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/service/WalletServiceImpl.java]: Done, changed used method from collect() to toList()

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java (hunk 6 above)
Comment (pyanik) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java]: Done, try/catch section has been moved to service layer. Proper log has been added.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java (hunk 7 above)
Comment (pyanik) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java]: Done, name of the attribute has been removed from annotation, only the passed  value remained. Thanks.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java (hunk 8 above)
Comment (pyanik) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java]: Not sure I understand your comment correctly, You mean that I should change the method signature of service method to return ResponseEntity object?

In my opinion as the ResponseEntity represents an HTTP response (which may consists of response status, message and body), so it is more related to Controller-layer where we prepare and send business output to the view.
It shouldn't be returned by Service-layer where we only handle the business logic. In our case, it is to return a list of all available wallets, which is then used as a body of our http response.

Could you provide an example of your proposed approach of creating a Controller-layer in the projects (Github projects/blog posts etc)?

**Review by Norsu296 — CHANGES_REQUESTED** [review]

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java (hunk 8 above)
Comment (Norsu296) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java]: Http response should be in controller layer, agree, but why u use response entity? If wallet is empty just return empty list and status ok, frontend know what to do with it.

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java [hunk 9]
```diff
@@ -14,4 +20,19 @@
...
     private final WalletService walletService;
+
+    @GetMapping("/wallets")
+    ResponseEntity<List<WalletDTO>> getWallets() {
+
+        List<WalletDTO> allWallets = walletService.getAllWallets();
+        HttpHeaders header =  new HttpHeaders();
+
+        if (!allWallets.isEmpty()) {
+            header.add("message", "The list of wallets has been successfully retrieved");
```
[the remark this answers is not in the document]
Comment (pyanik) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java:31]: According to issue description we have to result the text of the operation.

Why we need to check the length of the list being sent. What do you mean by that?

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java [hunk 10]
```diff
@@ -14,4 +20,19 @@
...
+        } else {
+            header.add("message", "There are no available wallets to view");
+        }
+
+        return ResponseEntity.ok().headers(header).body(allWallets);
```
[the remark this answers is not in the document]
Comment (pyanik) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java:36]: In othercase I can not use static builder of ResponseEntity class

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/exception/GlobalExceptionHandlerController.java [hunk 11]
```diff
@@ -35,4 +33,11 @@ public String catchAppRuntimeException(AppRuntimeException e) {
         log.error("AppRuntimeException", e);
         return e.getMessage();
     }
+
+    @ExceptionHandler(ListOfWalletsNotFoundException.class)
```
[the remark this answers is not in the document]
Comment (pyanik) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/exception/GlobalExceptionHandlerController.java]: Good point, ListOfWalletsNotFoundException has been replaced with ResourceNotFoundException class

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/exception/ListOfWalletsNotFoundException.java [hunk 12]
```diff
@@ -0,0 +1,11 @@
+package pl.byczazagroda.trackexpensesappbackend.exception;
+
+/**
+ * WalletsNotFoundException business exception {@code AppRuntimeException}.
+ */
+public class ListOfWalletsNotFoundException extends AppRuntimeException {
+
+    public ListOfWalletsNotFoundException() {
+        super("An error occurred while retrieving the list of wallets");
```
[the remark this answers is not in the document]
Comment (pyanik) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/exception/ListOfWalletsNotFoundException.java]: One parameter constructor for ResourceNotFoundException with String value has been added

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java [hunk 13]
```diff
@@ -14,4 +20,19 @@
 public class WalletController {
 
     private final WalletService walletService;
+
+    @GetMapping("/wallets")
```
[the remark this answers is not in the document]
Comment (pyanik) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java]: Done, now endpoint for get is http://url/api/wallet

But is seems there is a discussion need for endpoints design in the project

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java (hunk 8 above)
Comment (pyanik) [reply src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/WalletController.java]: Given  get method not only handle the situation when empty list is returned

**Pull request closed without merging**